src/vm/costs/mod.rs

```diff
@@ -194,6 +260,235 @@ pub enum CostErrors {
+            .expect_tuple();
+        let target_contract = match entry
+            .get("function-contract")
+            .unwrap()
+            .clone()
+            .expect_principal()
+        {
+            PrincipalData::Contract(contract_id) => contract_id,
+            _ => {
+                warn!("Confirmed cost proposal invalid: function-contract is not a contract principal";
+                          "confirmed_proposal_id" => confirmed_proposal);
+                continue;
+            }
+        };
+        let target_function = match ClarityName::try_from(
+            entry.get("function-name").unwrap().clone().expect_ascii(),
+        ) {
+            Ok(x) => x,
+            Err(_) => {
+                warn!("Confirmed cost proposal invalid: function-name is not a valid function name";
+                          "confirmed_proposal_id" => confirmed_proposal);
+                continue;
+            }
+        };
+        let cost_contract = match entry
+            .get("cost-function-contract")
+            .unwrap()
+            .clone()
+            .expect_principal()
+        {
+            PrincipalData::Contract(contract_id) => contract_id,
+            _ => {
+                warn!("Confirmed cost proposal invalid: cost-function-contract is not a contract principal";
+                          "confirmed_proposal_id" => confirmed_proposal);
+                continue;
+            }
+        };
+
+        let cost_function = match ClarityName::try_from(
+            entry
+                .get_owned("cost-function-name")
+                .unwrap()
+                .expect_ascii(),
+        ) {
+            Ok(x) => x,
+            Err(_) => {
+                warn!("Confirmed cost proposal invalid: cost-function-name is not a valid function name";
+                          "confirmed_proposal_id" => confirmed_proposal);
+                continue;
+            }
+        };
+
+        // Here is where we perform the required validity checks for a confirmed proposal:
+        //  * Replaced contract-calls _must_ be `define-read-only` _or_ refer to one of the boot code
+        //      cost functions
+        //  * cost-function contracts must be arithmetic only
+
+        // make sure the contract is "cost contract eligible" via the
+        //  arithmetic-checking analysis pass
+        let cost_func_ref = match clarity_db.load_contract_analysis(&cost_contract) {
+            Some(c) => {
+                if !c.is_cost_contract_eligible {
+                    warn!("Confirmed cost proposal invalid: cost-function-contract uses non-arithmetic or otherwise illegal operations";
+                          "confirmed_proposal_id" => confirmed_proposal,
+                          "contract_name" => %cost_contract,
+                    );
+                    continue;
+                }
+                if !c.public_function_types.contains_key(&cost_function)
```

Comment:
We don't care if a cost function is `define-public`, `define-read-only`, or `define-private`?  Would it help things any if we constrained cost functions to all be `define-read-only`?  I'm not sure a cost function could be `define-public`, for example, if the cost function wasn't already required to evaluate to a `(response A B)`.

Reply:
More specifically, `cost_contract` is the contract that contains the cost function, right?  If so, I think you may have confused this `if`-condition with the `if`-condition below for `target_contract` (which I think is the contract who is about to get the new cost function?)

Reply:
No, these `if` statements are correct, the`target` should always be a read-only function. The cost function can only be a read-only or private (though this `if` statement shouldn't check public functions, which wouldn't be allowed in any case by the cost-contract eligibility analysis).

Reply:
The `public` option has been removed in https://github.com/blockstack/stacks-blockchain/pull/2143/commits/c03277838d1e045d9f6c72c2e13e861b90819369